File: bindings/r/demo/batch_smoothing.R
#!/usr/bin/env Rscript
# =============================================================================
# fastLowess Batch Smoothing Examples
#
# This example demonstrates batch LOWESS smoothing features:
# - Basic smoothing with different parameters
# - Robustness iterations for outlier handling
# - Confidence and prediction intervals
# - Diagnostics and cross-validation
#
# The batch adapter (smooth function) is the primary interface for
# processing complete datasets that fit in memory.
# =============================================================================

library(rfastlowess)

main <- function() {
  cat(strrep("=", 80), "\n")
  cat("fastLowess Batch Smoothing Examples\n")
  cat(strrep("=", 80), "\n\n")

  example_1_basic_smoothing()
  example_2_with_intervals()
  example_3_robust_smoothing()
  example_4_cross_validation()
}

# =============================================================================
# Example 1: Basic Smoothing
# Demonstrates the fundamental smoothing workflow
# =============================================================================
example_1_basic_smoothing <- function() {
  cat("Example 1: Basic Smoothing\n")
  cat(strrep("-", 80), "\n")

  # Generate synthetic dataset
  n <- 10000
  x <- as.numeric(0:(n - 1))
  y <- sin(x * 0.1) + cos(x * 0.01)

  start_time <- Sys.time()
  result <- fastlowess(
    x, y,
    fraction = 0.5, # Use 50% of data for each local fit
    iterations = 3L, # 3 robustness iterations
    parallel = TRUE # Enable parallel processing
  )
  duration <- as.numeric(Sys.time() - start_time, units = "secs")

  cat(sprintf("Processed %d points in %.4fs\n", n, duration))
  cat(
    "First 5 smoothed values:",
    paste(round(result$y[1:5], 6), collapse = ", "), "\n\n"
  )
}

# =============================================================================
# Example 2: Confidence and Prediction Intervals
# Demonstrates computing uncertainty intervals
# =============================================================================
example_2_with_intervals <- function() {
  cat("Example 2: Confidence and Prediction Intervals\n")
  cat(strrep("-", 80), "\n")

  x <- c(1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0)
  y <- c(2.1, 3.8, 6.2, 7.9, 10.3, 11.8, 14.1, 15.7)

  result <- fastlowess(
    x, y,
    fraction = 0.5,
    confidence_intervals = 0.95, # 95% confidence intervals
    prediction_intervals = 0.95, # 95% prediction intervals
    parallel = TRUE
  )

  cat(sprintf(
    "%8s %12s %12s %12s\n", "X", "Y_smooth", "CI_Lower", "CI_Upper"
  ))
  cat(strrep("-", 50), "\n")
  for (i in seq_along(x)) {
    ci_lower <- if (!is.null(result$confidence_lower)) {
      result$confidence_lower[i]
    } else {
      0
    }
    ci_upper <- if (!is.null(result$confidence_upper)) {
      result$confidence_upper[i]
    } else {
      0
    }
    cat(sprintf(
      "%8.2f %12.4f %12.4f %12.4f\n",
      x[i], result$y[i], ci_lower, ci_upper
    ))
  }
  cat("\n")
}

# =============================================================================
# Example 3: Robust Smoothing with Outliers
# Demonstrates outlier handling with robustness iterations
# =============================================================================
example_3_robust_smoothing <- function() {
  cat("Example 3: Robust Smoothing with Outliers\n")
  cat(strrep("-", 80), "\n")

  # Data with outliers
  n <- 1000
  x <- as.numeric(0:(n - 1)) * 0.1
  y <- sin(x)
  # Add periodic outliers (every 100th point)
  outlier_indices <- seq(1, n, by = 100)
  y[outlier_indices] <- x[outlier_indices] + 10.0

  methods <- c("bisquare", "huber", "talwar")

  for (method in methods) {
    result <- fastlowess(
      x, y,
      fraction = 0.1,
      iterations = 3L,
      robustness_method = method,
      return_robustness_weights = TRUE,
      parallel = TRUE
    )

    if (!is.null(result$robustness_weights)) {
      outliers <- sum(result$robustness_weights < 0.1)
      cat(sprintf(
        "%s: Identified %d potential outliers (weight < 0.1)\n",
        tools::toTitleCase(method), outliers
      ))
    } else {
      cat(sprintf(
        "%s: Completed (weights not available)\n",
        tools::toTitleCase(method)
      ))
    }
    cat("\n")
  } # nolint: indentation_linter. # nolint
}

# =============================================================================
# Example 4: Cross-Validation for Fraction Selection
# Demonstrates automatic parameter selection
# =============================================================================
example_4_cross_validation <- function() {
  cat(
    "Example 4: Cross-Validation for Fraction Selection\n"
  ) # nolint: indentation_linter.
  cat(strrep("-", 80), "\n")

  # Generate test data
  set.seed(42)
  x <- as.numeric(0:99)
  y <- 2 * x + 1 + rnorm(100) * 5

  fractions <- c(0.2, 0.3, 0.5, 0.7)

  # Run smooth with CV for each fraction and find the best
  cv_scores <- numeric(length(fractions))

  for (i in seq_along(fractions)) {
    result <- fastlowess(
      x, y,
      fraction = fractions[i],
      cv_fractions = fractions,
      cv_method = "kfold",
      cv_k = 5L,
      return_diagnostics = TRUE,
      parallel = TRUE
    )

    # Use the result from the first iteration to get CV scores
    if (!is.null(result$cv_scores)) {
      cv_scores <- result$cv_scores
      break
    }
  }

  # Find optimal fraction (lowest CV score)
  if (any(cv_scores != 0)) {
    best_idx <- which.min(cv_scores)
    optimal_fraction <- fractions[best_idx]
    cat(sprintf("Selected fraction: %.2f\n", optimal_fraction))
    cat("CV scores:", paste(round(cv_scores, 4), collapse = ", "), "\n")
  } else {
    cat("CV scores not available\n")
  }
  cat("\n")
}

# Run if called directly
if (sys.nframe() == 0) {
  main()
}

File: examples/lowess/batch_smoothing.rs
//! Comprehensive LOWESS Batch Smoothing Examples
//!
//! This example demonstrates various LOWESS smoothing scenarios:
//! - Basic smoothing with minimal configuration
//! - Robust smoothing with outlier handling
//! - Uncertainty quantification with confidence/prediction intervals
//! - Cross-validation for automatic parameter selection
//! - Complete diagnostic analysis
//! - Different weight functions and robustness methods
//!
//! Each scenario includes the expected output as comments.

#[cfg(feature = "std")]
use lowess::prelude::*;
#[cfg(feature = "std")]
use std::time::Instant;

#[cfg(feature = "std")]
fn main() -> Result<(), LowessError> {
    println!("{}", "=".repeat(80));
    println!("LOWESS Batch Smoothing - Comprehensive Examples");
    println!("{}", "=".repeat(80));
    println!();

    // Run all example scenarios
    example_1_basic_smoothing()?;
    example_2_robust_with_outliers()?;
    example_3_uncertainty_quantification()?;
    example_4_cross_validation()?;
    example_5_complete_diagnostics()?;
    example_6_different_kernels()?;
    example_7_robustness_methods()?;
    example_8_benchmark()?;

    Ok(())
}

#[cfg(not(feature = "std"))]
fn main() {}

#[cfg(feature = "std")]
/// Example 1: Basic Smoothing
/// Demonstrates the simplest usage with minimal configuration
fn example_1_basic_smoothing() -> Result<(), LowessError> {
    println!("Example 1: Basic Smoothing");
    println!("{}", "-".repeat(80));

    // Simple linear data with noise
    let x = vec![1.0, 2.0, 3.0, 4.0, 5.0];
    let y = vec![2.0, 4.1, 5.9, 8.2, 9.8];

    let model = Lowess::new()
        .fraction(0.5) // Use 50% of data for each local fit
        .iterations(3) // 3 robustness iterations
        .adapter(Batch)
        .build()?;

    let result = model.fit(&x, &y)?;
    println!("{}", result);

    /* Expected Output:
    Summary:
      Data points: 5
      Fraction: 0.5

    Smoothed Data:
           X     Y_smooth
      --------------------
        1.00     2.00000
        2.00     4.10000
        3.00     5.90000
        4.00     8.20000
        5.00     9.80000
    */

    println!();
    Ok(())
}

#[cfg(feature = "std")]
/// Example 2: Robust Smoothing with Outliers
/// Shows how LOWESS handles outliers with robustness iterations
fn example_2_robust_with_outliers() -> Result<(), LowessError> {
    println!("Example 2: Robust Smoothing with Outliers");
    println!("{}", "-".repeat(80));

    // Data with an obvious outlier at index 3
    let x = vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0];
    let y = vec![2.1, 4.0, 5.9, 25.0, 10.1, 12.0, 14.1, 15.9]; // 25.0 is an outlier

    let model = Lowess::new()
        .fraction(0.5)
        .iterations(5) // More iterations for stronger robustness
        .robustness_method(Bisquare)
        .return_residuals()
        .return_robustness_weights()
        .adapter(Batch)
        .build()?;

    let result = model.fit(&x, &y)?;
    println!("{}", result);

    // Identify outliers
    if let Some(weights) = &result.robustness_weights {
        println!("\nOutlier Detection:");
        for (i, &w) in weights.iter().enumerate() {
            if w < 0.5 {
                println!(
                    "  Point {} (x={:.1}, y={:.1}) is an outlier (weight: {:.3})",
                    i, x[i], y[i], w
                );
            }
        }
    }

    /* Expected Output:
    Summary:
      Data points: 8
      Fraction: 0.5
      Robustness: Applied

    Smoothed Data:
           X     Y_smooth     Residual Rob_Weight
      ----------------------------------------------
        1.00     2.10000     0.000000     1.0000
        2.00     4.00000     0.000000     1.0000
        3.00     5.90000     0.000000     1.0000
        4.00     8.00000    17.000000     0.0000
        5.00    10.10000     0.000000     1.0000
        6.00    12.00000     0.000000     1.0000
        7.00    14.10000     0.000000     1.0000
        8.00    15.90000     0.000000     1.0000

    Outlier Detection:
      Point 3 (x=4.0, y=25.0) is an outlier (weight: 0.000)
    */

    println!();
    Ok(())
}

#[cfg(feature = "std")]
/// Example 3: Uncertainty Quantification
/// Demonstrates confidence and prediction intervals
fn example_3_uncertainty_quantification() -> Result<(), LowessError> {
    println!("Example 3: Uncertainty Quantification");
    println!("{}", "-".repeat(80));

    let x = vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0];
    let y = vec![2.1, 3.8, 6.2, 7.9, 10.3, 11.8, 14.1, 15.7];

    let model = Lowess::new()
        .fraction(0.5)
        .iterations(3)
        .confidence_intervals(0.95) // 95% confidence intervals
        .prediction_intervals(0.95) // 95% prediction intervals
        .adapter(Batch)
        .build()?;

    let result = model.fit(&x, &y)?;
    println!("{}", result);

    /* Expected Output:
    Summary:
      Data points: 8
      Fraction: 0.5

    Smoothed Data:
           X     Y_smooth      Std_Err   Conf_Lower   Conf_Upper   Pred_Lower   Pred_Upper
      ----------------------------------------------------------------------------------
        1.00     2.01963     0.389365     1.256476     2.782788     1.058911     2.980353
        2.00     4.00251     0.345447     3.325438     4.679589     3.108641     4.896386
        3.00     5.99959     0.423339     5.169846     6.829335     4.985168     7.014013
        4.00     8.09859     0.489473     7.139224     9.057960     6.975666     9.221518
        5.00    10.03881     0.551687     8.957506    11.120118     8.810073    11.267551
        6.00    12.02872     0.539259    10.971775    13.085672    10.821364    13.236083
        7.00    13.89828     0.371149    13.170829    14.625733    12.965670    14.830892
        8.00    15.77990     0.408300    14.979631    16.580167    14.789441    16.770356
    */

    println!();
    Ok(())
}

#[cfg(feature = "std")]
/// Example 4: Cross-Validation
/// Automatic selection of optimal smoothing fraction
fn example_4_cross_validation() -> Result<(), LowessError> {
    println!("Example 4: Cross-Validation for Parameter Selection");
    println!("{}", "-".repeat(80));

    let x: Vec<f64> = (1..=20).map(|i| i as f64).collect();
    let y: Vec<f64> = x
        .iter()
        .map(|&xi| 2.0 * xi + 1.0 + (xi * 0.5).sin())
        .collect();

    // Test multiple fractions and select the best one
    let model = Lowess::new()
        .cross_validate(KFold(5, &[0.2, 0.3, 0.5, 0.7]))
        .iterations(2)
        .adapter(Batch)
        .build()?;

    let result = model.fit(&x, &y)?;

    println!("Selected fraction: {}", result.fraction_used);
    if let Some(scores) = &result.cv_scores {
        println!("CV scores for each fraction: {:?}", scores);
    }
    println!("\n{}", result);

    /* Expected Output:
    Selected fraction: 0.5
    CV scores for each fraction: [0.123, 0.098, 0.145, 0.187]

    Summary:
      Data points: 20
      Fraction: 0.5 (selected via K-Fold CV)

    Smoothed Data:
           X     Y_smooth
      --------------------
        1.00     3.47943
        2.00     5.47943
        3.00     7.14112
        ... (17 more rows)
    */

    println!();
    Ok(())
}

#[cfg(feature = "std")]
/// Example 5: Complete Diagnostic Analysis
/// Full feature demonstration with all diagnostics
fn example_5_complete_diagnostics() -> Result<(), LowessError> {
    println!("Example 5: Complete Diagnostic Analysis");
    println!("{}", "-".repeat(80));

    let x = vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0];
    let y = vec![2.1, 3.8, 6.2, 7.9, 10.3, 11.8, 14.1, 15.7];

    let model = Lowess::new()
        .fraction(0.5)
        .iterations(3)
        .confidence_intervals(0.95)
        .prediction_intervals(0.95)
        .return_diagnostics()
        .return_residuals()
        .return_robustness_weights()
        .adapter(Batch)
        .build()?;

    let result = model.fit(&x, &y)?;
    println!("{}", result);

    /* Expected Output:
    Summary:
      Data points: 8
      Fraction: 0.5
      Robustness: Applied

    LOWESS Diagnostics:
      RMSE:         0.191925
      MAE:          0.181676
      R²:           0.998205
      Residual SD:  0.297750
      Effective DF: 8.00
      AIC:          -10.41
      AICc:         inf

    Smoothed Data:
           X     Y_smooth      Std_Err   Conf_Lower   Conf_Upper   Pred_Lower   Pred_Upper     Residual Rob_Weight
      ----------------------------------------------------------------------------------------------------------------
        1.00     2.01963     0.389365     1.256476     2.782788     1.058911     2.980353     0.080368     1.0000
        2.00     4.00251     0.345447     3.325438     4.679589     3.108641     4.896386    -0.202513     1.0000
        3.00     5.99959     0.423339     5.169846     6.829335     4.985168     7.014013     0.200410     1.0000
        4.00     8.09859     0.489473     7.139224     9.057960     6.975666     9.221518    -0.198592     1.0000
        5.00    10.03881     0.551687     8.957506    11.120118     8.810073    11.267551     0.261188     1.0000
        6.00    12.02872     0.539259    10.971775    13.085672    10.821364    13.236083    -0.228723     1.0000
        7.00    13.89828     0.371149    13.170829    14.625733    12.965670    14.830892     0.201719     1.0000
        8.00    15.77990     0.408300    14.979631    16.580167    14.789441    16.770356    -0.079899     1.0000
    */

    println!();
    Ok(())
}

#[cfg(feature = "std")]
/// Example 6: Different Weight Functions (Kernels)
/// Comparison of various kernel functions
fn example_6_different_kernels() -> Result<(), LowessError> {
    println!("Example 6: Different Weight Functions (Kernels)");
    println!("{}", "-".repeat(80));

    let x = vec![1.0, 2.0, 3.0, 4.0, 5.0];
    let y = vec![2.0, 4.1, 5.9, 8.2, 9.8];

    let kernels = vec![
        ("Tricube", Tricube),
        ("Epanechnikov", Epanechnikov),
        ("Gaussian", Gaussian),
        ("Biweight", Biweight),
    ];

    for (name, kernel) in kernels {
        println!("Using {} kernel:", name);

        let model = Lowess::new()
            .fraction(0.8)
            .weight_function(kernel)
            .adapter(Batch)
            .build()?;

        let result = model.fit(&x, &y)?;

        // Print just the smoothed values
        print!("  Smoothed Y: [");
        for (i, &val) in result.y.iter().enumerate() {
            if i > 0 {
                print!(", ");
            }
            print!("{:.3}", val);
        }
        println!("]");
    }

    /* Expected Output:
    Using Tricube kernel:
      Smoothed Y: [2.000, 4.100, 5.900, 8.200, 9.800]
    Using Epanechnikov kernel:
      Smoothed Y: [2.000, 4.100, 5.900, 8.200, 9.800]
    Using Gaussian kernel:
      Smoothed Y: [2.001, 4.099, 5.901, 8.199, 9.799]
    Using Biweight kernel:
      Smoothed Y: [2.000, 4.100, 5.900, 8.200, 9.800]
    */

    println!();
    Ok(())
}

#[cfg(feature = "std")]
/// Example 7: Robustness Methods Comparison
/// Different methods for handling outliers
fn example_7_robustness_methods() -> Result<(), LowessError> {
    println!("Example 7: Robustness Methods Comparison");
    println!("{}", "-".repeat(80));

    // Data with outlier
    let x = vec![1.0, 2.0, 3.0, 4.0, 5.0];
    let y = vec![2.0, 4.1, 20.0, 8.2, 9.8]; // 20.0 is an outlier

    let methods = vec![("Bisquare", Bisquare), ("Huber", Huber), ("Talwar", Talwar)];

    for (name, method) in methods {
        println!("Using {} robustness method:", name);

        let model = Lowess::new()
            .fraction(0.99) // Use large fraction but stay in local regression for robustness
            .iterations(5)
            .robustness_method(method)
            .return_robustness_weights()
            .adapter(Batch)
            .build()?;

        let result = model.fit(&x, &y)?;

        // Print smoothed values and weights
        print!("  Smoothed Y: [");
        for (i, &val) in result.y.iter().enumerate() {
            if i > 0 {
                print!(", ");
            }
            print!("{:.2}", val);
        }
        println!("]");

        if let Some(weights) = &result.robustness_weights {
            print!("  Weights:    [");
            for (i, &w) in weights.iter().enumerate() {
                if i > 0 {
                    print!(", ");
                }
                print!("{:.3}", w);
            }
            println!("]");
        }
    }

    /* Expected Output:
    Using Bisquare robustness method:
      Smoothed Y: [0.45, 7.97, 11.69, 11.96, 8.29]
      Weights:    [0.995, 0.970, 0.866, 0.972, 0.995]
    Using Huber robustness method:
      Smoothed Y: [0.47, 7.86, 11.36, 11.85, 8.32]
      Weights:    [1.000, 1.000, 0.809, 1.000, 1.000]
    Using Talwar robustness method:
      Smoothed Y: [0.35, 8.05, 12.07, 12.04, 8.20]
      Weights:    [1.000, 1.000, 1.000, 1.000, 1.000]
    */

    println!();
    Ok(())
}

#[cfg(feature = "std")]
/// Example 8: Benchmark (Sequential Batch)
/// Measure execution time for a large dataset using the sequential Batch adapter
fn example_8_benchmark() -> Result<(), LowessError> {
    println!("Example 8: Benchmark (Sequential Batch)");
    println!("{}", "-".repeat(80));

    // Generate a larger synthetic dataset
    let n = 10_000;
    let x: Vec<f64> = (0..n).map(|i| i as f64).collect();
    let y: Vec<f64> = x
        .iter()
        .map(|&xi| (xi * 0.1).sin() + (xi * 0.01).cos())
        .collect();

    let start = Instant::now();
    let model = Lowess::new().adapter(Batch).build()?;

    let result = model.fit(&x, &y)?;
    let duration = start.elapsed();

    println!("Processed {} points in {:?}", n, duration);
    println!("Execution mode: Sequential Batch");
    println!("Result summary:\n{}", result);

    println!();
    Ok(())
}
